### PR TITLE
Add baseurl for assets

### PR DESCRIPTION
**Pull request by vis-kid**
Assets now stay intact on deployed gh-pages branch from fork
Add `baseurl: /homepage` to config file (user/org.github.io/homepage)
Add baseurl to style asset link in head tag
Add baseurl to image asset paths

### CODE REVIEW

File: _config.yml [hunk 1]
```diff
@@ -5,3 +5,5 @@ exclude:
 
 gems:
   - jekyll-redirect-from
+
+baseurl: /homepage
```
Comment (rafaelfranca) [inline _config.yml]: Would not this break this project? If I add `/homepage` in the currently generated page it doesn't work.

**Comment by vis-kid** [comment]
I had the same concern but since under the hood `rubyonrails.org` is served as `rails.github.io/homepage` I assumed it is safe. 

What exactly did not work?

**Comment by rafaelfranca** [comment]
`http://rubyonrails.org/homepage/style.css` give 404 to me, and it is this URL that is generated.

**Comment by vis-kid** [comment]
Not sure…

Does `http://rubyonrails.org/style.css` work then?

On my end, `http://rails.github.io/homepage/style.css` works just fine and gets translated into `http://rubyonrails.org/style.css`. 😲 💭 

When `http://rails.github.io/homepage` translates to `http://rubyonrails.org` it should do the trick, no? 😄 

If I missed something, I’m not sure what it is. But it’s almost 5am so I’m gonna look at this with fresh eyes after some shuteye…

**Comment by rafaelfranca** [comment]
> Does http://rubyonrails.org/style.css work then?

Yes, it does.



> When http://rails.github.io/homepage translates to http://rubyonrails.org it should do the trick, no? 😄

No, it is not translated. The URL generated in the browser is `/homepage/style.css` and this tell the browser to download that file in the current domain that is `http://rubyonrails.org`.

**Comment by vis-kid** [comment]
What confuses me is that I did the exact same thing on the Sinatra site recently and it worked effortlessly. I’ll investigate with fresh eyes :)

**Comment by rafaelfranca** [comment]
Weird. Let me try a live test 😄

**Comment by rafaelfranca** [comment]
Yeah, it didn't work indeed. See the page:

screen shot 2016-08-23 at 1 18 19 am

And the source code:

```
<!DOCTYPE html>
<html>
  <head>
    <title>Ruby on Rails</title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">
    <link rel="stylesheet" href="/homepage/style.css" type="text/css" media="screen" charset="utf-8">
  </head>

  <body>
    <div class="container">
          <div class="nav">
      <ul>
        <li class="first"><a href="http://weblog.rubyonrails.org/">Blog</a></li>
        <li><a href="http://guides.rubyonrails.org/">Guides</a></li>
        <li><a href="http://api.rubyonrails.org/">API</a></li>
        <li><a href="http://stackoverflow.com/questions/tagged/ruby-on-rails">Ask for help</a></li>
        <li><a href="https://github.com/rails/rails">Contribute on GitHub</a></li>
      </ul>
    </div>

    <section>
      <p class="mobile-center">
      <img src="/homepage/images/rails-logo.svg" width="220" height="78" alt="Ruby on Rails">
      </p>
    </section>

    <section>
      <figure class="right">
      <img src="/homepage/images/imagine.png" alt="So many possibilities.">
      </figure>
```

**Comment by vis-kid** [comment]
Thanks for the live test! :metal: 

Checked the Sinatra site. No biggie there. :)

With `baseurl: /sinatra.github.com`, the generated link 

`http://www.sinatrarb.com/sinatra.github.com/css/application.css` 

gets translated into

 `http://www.sinatrarb.com/css/application.css`.

Bug or me sitting in front of a screen for over 14hrs.

Hmm… 🎻 💭

**Comment by vis-kid** [comment]
@rafaelfranca 

Hi! Sorry to bother you again with this.

Well, an easy “fix” would be to define a `url: rubyonrails.org` in config and use `site.url` instead. I wanted to avoid that and provide other contributors an easy time to push their changes to their gh-pages branch for inspection—without the need to adjust this setting on their end all the time. Since this site is developed against the gh-pages branch directly, I think `baseurl` would be the smoothest option in that regard.

Since I did exactly the same as on the Sinatra site, I’m not sure I’m causing this. I have two questions: Can we preempt the possibilities that` exlude -CNAME` in `_config.yml` or developing on gh-pages directly has any influence on this issue?

**Comment by rafaelfranca** [comment]
You are not bothering at all.

Yeah, `url` seems to work but I agree with you that it is more work to contributors.

I'm not sure why `CNAME` is in the `exclude` option, but I believe the cause of this is the development being in `gh-pages`. At least this is  the only difference that I see between Sinatra's site and Rails' site.

**Comment by vis-kid** [comment]
My thoughts exactly :)

Would like to know though if excluding CNAME is possibly obsolete…

What do you recommend we should do now? Is rebasing / moving development back to master an option at all?  And do you know why it was moved from master to gh-pages back in March?

**Comment by rafaelfranca** [comment]
I believe because master only work if the repository name is the same as the organization name with the suffix `github.io`. If it works with a different name we can change back to the master branch without any problem.

**Comment by vis-kid** [comment]
https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/
https://help.github.com/articles/user-organization-and-project-pages/

Bottom line:

screen shot 2016-08-23 at 10 31 32 pm


### GH Pages docs:
> If your site is a User or Organization Page that has a repository named <username>.github.io or <orgname>.github.io, you cannot publish your site's source files from different locations. User and Organization Pages that have this type of repository name are only published from the master branch.

Project pages on the other hand have both the `master` and `gh-pages` branches as an option. So I guess that means in any case `master` is a safe bet.

If you switch back we might catch two birds with one stone:

- Chances are pretty good `baseurl` might work out of the box like the Sinatra site.
- Contributors can also push and build their own `gh-pages` branch for inspection more smoothly if we decide to go with `site.url` instead. But in any scenario that I considered, without `baseurl` working, even that becomes a pain the neck over time…

**Comment by vis-kid** [comment]
Any updates on this? :)

I know this can’t be super high on your to-do list but figuring this out first would make my design suggestions a lot more straightforward to implement and show…

P.S. Apologies if my etiquette is not finely tuned yet. Still new to the OSS party. 😄

**Comment by rafaelfranca** [comment]
Oh, sorry. Yeah we can change it back to master and see how it works. I can't do it right now, but I'll try to do sometime this week.

**Comment by vis-kid** [comment]
Sure. Got it. :)

**Comment by vis-kid** [comment]
@rafaelfranca Hi! If you are not too busy, do you think you’ll have time for this issue this week? :)

**Comment by rafaelfranca** [comment]
Oh, sorry I'm still busy this week but I feel bad for blocking your work. Let me try to get someone else from the core to do this.

**Comment by vis-kid** [comment]
No worries @rafaelfranca, really, it’s no biggie :)

**Comment by vis-kid** [comment]
I can wait until you are free if necessary. Fiddling around with deployment branches atm is fine for now. Just wanted to avoid putting future contributors—who are maybe more on the design side of things—through the same hoops. Whenever you are ready is cool with me…

**Comment by rafaelfranca** [comment]
I was wondering, are not contributors able to test their changes running jekyll locally?

**Comment by vis-kid** [comment]
Sure, that is fine, all working here. But to make the process for design changes a  lot more smooth, it is nice if the core team for example can immediately see the changes in action.

It’s one thing to look at the code and screenshots but nothing beats looking at a live site and playing around with the changes. Saves time for everybody I guess—discussions maybe too. :)

**Comment by vis-kid** [comment]
Plus, changes that should be checked on multiple devices are smoother to do without fiddling around of course. :)

**Comment by pixeltrix** [comment]
@rafaelfranca @vis-kid happy to help out with the changes but before we do we'd need master to be in shape - it's a 137 commits behind from what I can see.

**Comment by vis-kid** [comment]
Hi Andrew @pixeltrix! Thanks very much for taking the time! Can you walk me through what I’d need to do exactly? Rebasing this branch on top of master?

**Comment by pixeltrix** [comment]
@vis-kid I assume that first I need to make master match the gh-pages branch and then you can rebase your branch on master.

**Comment by vis-kid** [comment]
@pixeltrix got it, nothing changed then. :) Misunderstanding.

**Comment by pixeltrix** [comment]
@vis-kid done now - I think you might need to create a new PR though

**Comment by vis-kid** [comment]
@pixeltrix sure, no problem

**Comment by vis-kid** [comment]
@pixeltrix The only possible important difference that we could identify between this project and the Sinatra page (where `baseurl` is working w/ fuzz) was the default branch. Here it is `gh-pages` and on Sinatra `master`.

If I understand the Jekyll / GH Pages magic correctly, then we want GitHub to build against `gh-pages` as is, but change the default to `master` here. That would at least smoothen the design workflow. Hope I’m not completely off the rails with this. :)

**Comment by pixeltrix** [comment]
> If I understand the Jekyll / GH Pages magic correctly, then we want GitHub to build against gh-pages as is, but change the default to master here. That would at least smoothen the design workflow because contributors can keep their own gh-pages branch completely separate and deploy new changes for inspection on their own fork

Trying to work out what the workflow is meant to be:

1. Default branch is master
2. Contributors fork and create a branch off master
3. They make some commits on that branch
4. They merge into their gh-pages branch and test it
5. When happy they create a PR from their branch against our master
6. We review the PR and when happy merge into master
7. When we want to deploy a set of changes we merge master into our gh-pages branch

@vis-kid @rafaelfranca how does this sound? Are there any GitHub guidelines on the correct workflow?

My biggest problem with this is that the fork's gh-pages branch won't match our gh-pages branch so contributors will be having to constantly reset that branch on their forks won't they?

**Comment by vis-kid** [comment]
That seems pretty much correct in my assessment. :) Small things I noticed though:

.4 Fork contributors simply push their local development branch `whatever-new-feature` as `gh-pages` up to their `origin` which GH-Pages then in turn uses to build the site. That makes their `origin gh-pages` branch easy to throw away if it doesn’t need to be synced with the original one. They don’t even need a local version of `gh-pages` on their machine. Also, the testing part goes both ways.  On the GH-Pages site of the fork, the core team can quickly see the proposed changes in action and decide based on a live site if the suggestions are cool or not.

.7 Under `Settings`, you would set your `master` as the source for GitHub Pages ([Configuring a publishing source for GitHub Pages](https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/)). That way you don’t need to merge `master` into the `gh-pages` branch at all anymore. In fact, your repo can get rid of `gh-pages` completely—just like the [Sinatra site](https://github.com/sinatra/sinatra.github.com/branches) did. What I don’t know yet is when people fork this repo, if they need to change this setting back to `gh-pages` as source for their version of the GH Pages site.

.biggest problem
Agreed! The easiest fix I see for this PR is to fork the repo again once you are all set up with `master` on your end and open a new PR with my changes on top of that… How this whole thing will affect previous forks when they pull is beyond my current knowledge though.

Overall, publishing directly from `master` hopefully also solves the whole `baseurl` issue while new design changes can be quickly introduced through a smooth workflow for everybody. If the `baseurl` issue needs more detective work to figure out, we will at least have made the process for future (design) contributions a bit more streamlined.

I know this seems a bit overly involved at first but I think once this is implemented, it will be worth the trouble. But of course, in case I don’t see the wood for the trees here, if there is an even simpler solution to all of this, I’d be happy to know. :) Maybe you can think of somebody with a bit more Jekyll magic insights to ping for a fresh set of eyes…

Sorry for the verbosity here, but since this seems a bit more delicate, I tried to avoid tripping up over conciseness. 😄

**Comment by vis-kid** [comment]
Hi guys! @pixeltrix, @rafaelfranca 
Any updates if we can move forward with this? It’s blocking me a little bit 😊

**Comment by pixeltrix** [comment]
@vis-kid sorry, a busy couple of weeks at work. Will re-read your instructions and see if it all makes sense.

**Comment by vis-kid** [comment]
Cool. Glad it worked! 😄

**Pull request merged**